src/bsd.rs

```diff
@@ -285,13 +318,45 @@ fn if_index(remote: IpAddr) -> Result<u16> {
         if len < size_of::<rt_msghdr>() {
             return Err(default_err());
         }
-        let reply: rt_msghdr = buf.into();
+        let (reply, mut sa) = buf.split_at(size_of::<rt_msghdr>());
+        let reply: rt_msghdr = reply.into();
         if reply.rtm_version == query_version && reply.rtm_pid == pid && reply.rtm_seq == query_seq
         {
             // This is a reply to our query.
             return if reply.rtm_type == query_type {
                 // This is the reply we are looking for.
-                Ok(reply.rtm_index)
+                // Some BSDs let us get the interface index and MTU directly from the reply.
+                let mtu: Option<usize> = if reply.rtm_rmx.rmx_mtu != 0 {
+                    Some(
+                        reply
+                            .rtm_rmx
+                            .rmx_mtu
+                            .try_into()
+                            .map_err(|e: TryFromIntError| unlikely_err(e.to_string()))?,
+                    )
+                } else {
+                    None
+                };
+                if reply.rtm_index != 0 {
+                    // Some BSDs return the interface index directly.
+                    Ok((reply.rtm_index, mtu))
+                } else {
+                    // For others, we need to extract it from the sockaddrs.
+                    for i in 0..RTAX_MAX {
+                        if (reply.rtm_addrs & (1 << i)) != 0 {
+                            let saddr =
+                                unsafe { ptr::read_unaligned(sa.as_ptr().cast::<sockaddr>()) };
+                            if saddr.sa_family == AF_LINK {
+                                let sdl = unsafe {
+                                    ptr::read_unaligned(sa.as_ptr().cast::<sockaddr_dl>())
+                                };
+                                return Ok((sdl.sdl_index, mtu));
+                            }
+                            (_, sa) = sa.split_at(sockaddr_len(saddr.sa_family)?);
```

Comment:
Do I understand correctly that we don't need to `sa.split_at` when `(reply.rtm_addrs & (1 << i)) == 0`?

Reply:
Yes. There are only as many `sockaddr`s attached to the message as there are set bits in `reply.rtm_addrs` (the set bits signify which response a `sockaddr` for a given position is for. It's a **terrible** interface.

Reply:
:exploding_head: